File: 0528. Random Pick with Weight/Solution0528.cpp
#include <iostream>     // std::cout
#include <algorithm>    // std::upper_bound
#include <vector>       // std::vector
using namespace std;

class Solution0528 {
    vector<int> v;
public:
    Solution0528(vector<int>& w) {
        v.push_back(w[0]);
        for (int i = 1; i < w.size(); i++) {
            v.push_back(v[i-1]+w[i]);
        }
    }
    
    int pickIndex() {
        int n = rand() % v[v.size()-1];
        auto it = upper_bound(v.begin(), v.end(), n);
        return it - v.begin();
    }
};


// https://leetcode.com/problems/random-pick-with-weight/discuss/671632/C%2B%2B-simple-and-easy-solution-with-explanation
// http://c.biancheng.net/view/7527.html
// http://c.biancheng.net/view/7521.html

File: 0528. Random Pick with Weight/Solution0528.java
class Solution0528 { // 50 - 54x
    int[] w;
    public Solution(int[] w) {
        this.w = w;
        for (int i = 1; i < w.length; i++) {
            w[i] += w[i-1];
        }
    }
    
    public int pickIndex() {
        int rand = (int) (Math.random() * (w[w.length-1]));
        // find the first in w that > rand
        int left = 0, right = w.length - 1, mid = 0;
        while (left < right - 1) {
            mid = left + (right - left) / 2;
            if (w[mid] > rand) {
                 right = mid;
            } else {
               left = mid;
            }
        }
        if (w[left] > rand) {
            return left;
        }
        return right;
    }
}